finesse/hardware/st10_controller.py

```diff
@@ -0,0 +1,343 @@
+    current position.
+    """
+
+    STEPS_PER_ROTATION = 50800
+    """The total number of steps in one full rotation of the mirror."""
+
+    def __init__(self, serial: Serial) -> None:
+        """Create a new ST10Controller.
+
+        Args:
+            serial: The serial device to communicate with the ST10 controller
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Malformed message received from device
+        """
+        self.serial = serial
+
+        # Check that we are connecting to an ST10
+        self._check_device_id()
+
+        # In case the motor is still moving, stop it now
+        self.stop_moving()
+
+        # Move mirror to home position
+        self._home_and_reset()
+
+        super().__init__()
+
+    @staticmethod
+    def create(
+        port: str,
+        baudrate: int = 9600,
+        timeout: float = 1.0,
+        *serial_args: Any,
+        **serial_kwargs: Any,
+    ):
+        """Create a new ST10Controller with the specified serial device properties.
+
+        Args:
+            port: Serial port name
+            baudrate: Serial port baudrate
+            timeout: How long to wait for read operations (seconds)
+            serial_args: Extra arguments to Serial constructor
+            serial_kwargs: Extra keyword arguments to Serial constructor
+        """
+        if "write_timeout" not in serial_kwargs:
+            serial_kwargs["write_timeout"] = timeout
+
+        serial = Serial(port, baudrate, *serial_args, timeout=timeout, **serial_kwargs)
+        return ST10Controller(serial)
+
+    def __del__(self) -> None:
+        """Leave mirror facing downwards when finished.
+
+        This prevents dust accumulating.
+        """
+        try:
+            self.stop_moving()
+            self.move_to("nadir")
+        except Exception as e:
+            logging.error(f"Failed to reset mirror to downward position: {e}")
+
+    def _check_device_id(self) -> None:
+        """Check that the ID is the correct one for an ST10.
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: The device ID is not for an ST10
+        """
+        self._write("MV")
+        if self._read() != "107F024":
+            raise ST10ControllerError("Device ID indicates this is not an ST10")
+
+    def _get_input_status(self, index: int) -> bool:
+        """Read the status of the device's inputs.
+
+        The inputs to the controller include digital inputs, as well as other properties
+        like alarm status. The exact meaning seems to varies between boards.
+
+        Args:
+            index: Which boolean value in the input status array to check
+        """
```

Comment:
Maybe it's worth quoting something from the manual regarding "input status" here, to clarify what's being checked and thus why there's a comparison to the string "1"?

From the manual:
A closed input is represented by a “0” (zero), and an open input is represented by a “1” (one).